libs/utils/src/shared_retryable.rs

```diff
@@ -0,0 +1,647 @@
+    ///
+    /// This complication exists because on pageserver we cannot use `tokio::spawn` directly
+    /// at this time.
+    pub async fn try_restart<F, Fut, E2>(
+        &self,
+        retry_with: F,
+    ) -> (
+        impl std::future::Future<Output = Result<T, E2>> + Send + 'static,
+        Option<impl std::future::Future<Output = ()> + Send + 'static>,
+    )
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = Result<T, E1>> + Send + 'static,
+        E2: From<E1> + Send + 'static,
+        E2: From<RetriedTaskPanicked>,
+    {
+        use futures::future::Either;
+
+        match self.decide_to_retry_or_join(retry_with).await {
+            Ok(terminal) => (Either::Left(async move { terminal }), None),
+            Err((rx, maybe_fut)) => {
+                let recv = Self::make_oneshot_alike_receiver(rx);
+
+                (Either::Right(recv), maybe_fut)
+            }
+        }
+    }
+
+    /// Returns a Ok if the previous attempt had resulted in a terminal result. Err is returned
+    /// when an attempt can be joined and possibly needs to be spawned.
+    async fn decide_to_retry_or_join<F, Fut, E2>(
+        &self,
+        retry_with: F,
+    ) -> Result<
+        Result<T, E2>,
+        (
+            tokio::sync::broadcast::Receiver<Result<T, E1>>,
+            Option<impl std::future::Future<Output = ()> + Send + 'static>,
+        ),
+    >
+    where
+        F: FnOnce() -> Fut,
+        Fut: std::future::Future<Output = Result<T, E1>> + Send + 'static,
+        E2: From<E1>,
+        E2: From<RetriedTaskPanicked>,
+    {
+        let mut g = self.inner.lock().await;
+
+        let maybe_rx = match g.as_ref() {
+            Some(MaybeDone::Done(Ok(t))) => return Ok(Ok(t.to_owned())),
+            Some(MaybeDone::Done(Err(e))) if e.is_permanent() => {
+                return Ok(Err(E2::from(e.to_owned())))
+            }
+            Some(MaybeDone::Pending(weak)) => {
+                // failure to upgrade can mean only one thing: there was an unexpected
+                // panic which we consider as a transient retryable error.
+                weak.upgrade()
+            }
+            Some(MaybeDone::Done(Err(_retryable))) => None,
+            None => None,
+        };
+
+        let (strong, maybe_fut) = match maybe_rx {
+            Some(strong) => (strong, None),
+            None => {
+                // new attempt
+                // panic safety: invoke the factory before configuring the pending value
+                let fut = retry_with();
+
+                let (strong, fut) = self.make_run_and_complete(fut, &mut g);
+                (strong, Some(fut))
+            }
+        };
+
+        // important: the Arc<Receiver> is not held after unlocking
+        // important: we resubscribe before lock is released to be sure to get a message which
+        // is sent once receiver is dropped
+        let rx = strong.resubscribe();
+        drop(strong);
```

Comment:
It's not so easy to keep a hold of this in outer scope anymore due to splitting to methods. If this was `std::mem::forget(strong);` the following tests would fail:

<details>
<summary>Test output</summary>

```
--- shared_retryable::tests::restartable_until_permanent stdout ----
thread 'shared_retryable::tests::restartable_until_permanent' panicked at 'not yet implemented: really unlucky', libs/utils/src/shared_retryable.rs:479:48
thread 'shared_retryable::tests::restartable_until_permanent' panicked at 'assertion failed: matches!(res, Err(OuterError :: Unlucky))', libs/utils/src/shared_retryable.rs:488:9

---- shared_retryable::tests::cancelling_spawner_is_fine stdout ----
thread 'shared_retryable::tests::cancelling_spawner_is_fine' panicked at 'called `Result::unwrap()` on an `Err` value: AttemptPanicked', libs/utils/src/shared_retryable.rs:565:32

---- shared_retryable::tests::service_example stdout ----
thread 'shared_retryable::tests::service_example' panicked at 'expected this to be the only Arc<Receiver<V>>: broadcast::Receiver', libs/utils/src/shared_retryable.rs:374:37
thread 'shared_retryable::tests::service_example' panicked at 'called `Result::unwrap()` on an `Err` value: TaskPanicked', libs/utils/src/shared_retryable.rs:646:43
```

</details>